### PR TITLE
fix issue 16490 -  Usage of attributes in inline asm blocks is not documented

### PR DESCRIPTION
**Pull request by BBasile**

### CODE REVIEW

**Comment by BBasile** [comment]
[new version](http://dtest.dlang.io/artifact/website-63bfba08fec3a90d559571cfaa94b3ef5b695472-1897618868a4af8ee0463201ac1b8187/web/spec/iasm.html)

File: spec/iasm.dd [hunk 1]
```diff
@@ -26,6 +26,38 @@ $(SPEC_S D x86 Inline Assembler,
...
+$(GRAMMAR
+$(GNAME AsmStatement):
+    $(D asm) $(GLINK FunctionAttributes)$(OPT) $(D {) $(GLINK AsmInstructionList)$(OPT) $(D })
+
+    $(GNAME AsmInstructionList):
+        $(GLINK AsmInstruction) $(D ;)
+        $(GLINK AsmInstruction) $(D ;) $(I AsmInstructionList)
+    )
+
+    $(P Assembler instructions are located in an `asm` statement.
```
Comment (MetaLang) [inline spec/iasm.dd]: I think this statement should be a little stronger, e.g., "Assembler instructions **must** be in an `asm` statement."

**Comment by MetaLang** [comment]
I'll wait for a second opinion before merging.

**Comment by wilzbach** [comment]
Rebased to restart DAutoTest

File: spec/iasm.dd [hunk 2]
```diff
@@ -26,6 +26,38 @@ $(SPEC_S D x86 Inline Assembler,
...
+$(H2 $(LNAME2 asmstatements, Asm statement))
+
+$(GRAMMAR
+$(GNAME AsmStatement):
+    $(D asm) $(GLINK FunctionAttributes)$(OPT) $(D {) $(GLINK AsmInstructionList)$(OPT) $(D })
+
+    $(GNAME AsmInstructionList):
+        $(GLINK AsmInstruction) $(D ;)
+        $(GLINK AsmInstruction) $(D ;) $(I AsmInstructionList)
+    )
```
Comment (wilzbach) [inline spec/iasm.dd]: Shouldn't this have the same order as here?
https://dlang.org/spec/statement.html#asm

File: spec/iasm.dd [hunk 3]
```diff
@@ -26,6 +26,38 @@ $(SPEC_S D x86 Inline Assembler,
...
+
+    $(P Assembler instructions must be located inside an `asm` block.
+        Like functions, `asm` statements must be anotated with adequate function attributes to be compatible with the caller.
+        Asm statments attributes must be explicit, they are not infered.
```
Comment (wilzbach) [inline spec/iasm.dd]: - statements
- "explictly defined"

File: spec/iasm.dd [hunk 4]
```diff
@@ -26,6 +26,38 @@ $(SPEC_S D x86 Inline Assembler,
...
+$(GNAME AsmStatement):
+    $(D asm) $(GLINK FunctionAttributes)$(OPT) $(D {) $(GLINK AsmInstructionList)$(OPT) $(D })
+
+    $(GNAME AsmInstructionList):
+        $(GLINK AsmInstruction) $(D ;)
+        $(GLINK AsmInstruction) $(D ;) $(I AsmInstructionList)
+    )
+
+    $(P Assembler instructions must be located inside an `asm` block.
+        Like functions, `asm` statements must be anotated with adequate function attributes to be compatible with the caller.
```
Comment (wilzbach) [inline spec/iasm.dd]: This sounds weird. I would just remove the entire sentence.

File: spec/iasm.dd [hunk 5]
```diff
@@ -26,6 +26,38 @@ $(SPEC_S D x86 Inline Assembler,
...
+        Asm statments attributes must be explicit, they are not infered.
+    )
+
+---
+void func1() pure nothrow @safe @nogc
+{
+    asm pure nothrow @trusted @nogc
+    {}
```
Comment (wilzbach) [inline spec/iasm.dd]: How about actually using assembler statements here?

File: spec/iasm.dd [hunk 6]
```diff
@@ -26,6 +26,38 @@ $(SPEC_S D x86 Inline Assembler,
...
+}
+
+void func2() @safe @nogc
+{
+    asm @nogc // Error: asm statement is assumed to be @system - mark it with '@trusted' if it is not
+    {}
```
Comment (wilzbach) [inline spec/iasm.dd]: How about actually using assembler statements here?

File: spec/iasm.dd (hunk 2 above)
Comment (BBasile) [reply spec/iasm.dd]: Actually i've copied this from grammar.dd.

File: spec/iasm.dd (hunk 4 above)
Comment (BBasile) [reply spec/iasm.dd]: No, this is the point of this documentation change. I meant that asm blocks must be annotated, just like local functions, despite of not being local functions.

File: spec/iasm.dd (hunk 5 above)
Comment (BBasile) [reply spec/iasm.dd]: Assuming you meant instructions and not ~~statements~~.
i don't think that this is required. The scope of the documentation here is the asm statement, not the instructions.

**Comment by andralex** [comment]
@wilzbach I'll pull this, feel free to add to it. Thanks!

**Pull request merged**